src/serialization.rs

```diff
@@ -181,6 +181,116 @@ pub mod dependency_criteria {
     }
 }
 
+pub mod policy {
+    use super::*;
+    use crate::format::{
+        PackageName, PackagePolicyEntry, PackageVersion, Policy, PolicyEntry, VetVersion,
+    };
+    use serde::de::value::Error as SerdeError;
+
+    const VERSION_SEPARATOR: &str = ":";
+
+    #[derive(serde::Serialize, serde::Deserialize)]
+    #[serde(transparent)]
+    pub struct AllPolicies(SortedMap<Spanned<String>, PolicyEntry>);
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum FromAllPoliciesError {
+        #[error("cannot mix versioned and unversioned policies for {name}")]
+        MixedVersioning { name: PackageName },
+        #[error("more than one policy provided for {name}:{version}")]
+        Duplicate {
+            name: PackageName,
+            version: PackageVersion,
+        },
+        #[error("deserialization error")]
+        Deserialization {
+            #[from]
+            source: Spanned<SerdeError>,
+        },
+    }
+
+    impl std::convert::TryFrom<AllPolicies> for Policy {
```

Comment:
Yeah, to be honest the main reason I used the conversions rather than using the type directly is because I had already written lots of code which was using the current `Policy` struct. But I can change it to use the structure directly.

Reply:
I was about to do this, but it occurred to me that one disadvantage is that the types will no longer directly enforce the constraints, which obviously is remedied by `validate`, but that is only a partial remedy. Code that manipulates this type could accidentally produce invalid states. If we think that's a relatively low risk versus the improved parse error messages, then I will proceed with the change.

Reply:
We already have many places where it's possible for us to generate an invalid state unfortunately, so I'm not sure it makes that much of a difference. Overall our static validation story for internal state isn't that great, because everything operates directly on the data deserialized from the store rather than on some other representation which e.g. binds criteria in a consistent manner.

Reply:
Welllll... the potential for producing invalid state in other parts of the code isn't a great reason to continue that practice :) But also, as far as I can tell, the introduction of the conversion for serialization won't produce worse error messages? I should probably verify this rather than blindly making the statement, but the serialization should basically behave as if `AllPolicies` were used in line, and in both cases there's a manual validation step (`AllPolicides` does it during deserialization, `validate` method obviously does it after the fact). So that leads me to believe neither case will produce error messages which have more context than the other.

But I don't need to die on this hill, so if you would still prefer to not have the conversion and additional type I don't mind changing it.

Reply:
The reason why I think that doing it in `validate` will be better is because we can actually report span information by using a proper diagnostic type with source locations, whereas errors which occur when deserializing can't report precise location information due to the serde boundary requiring the errors be simplified to a string.

We can also include extra hints like "help" lines if we do custom errors from `validate()`, which is also not possible from within a serde deserializer.

I don't necessarily think it's the most important thing to do this sort of simplification, and if it'd be a lot of hassle for you, we definitely don't need to go down this route.

Reply:
That's fair. I was somewhat unfamiliar of whether, for instance, the diagnostic message generator was doing something "clever" with the Error types, though I guess it couldn't do that much (I recall seeing proposals for `Error: Any` for these sorts of things). I will change to a `validate()`.

Reply:
Yeah, more detailed error propagation would require https://rust-lang.github.io/rfcs/3192-dyno.html or something like it, and that's not stable yet (https://github.com/rust-lang/rust/issues/96024). The provider API would also generally be useful for other runtime serde customizations if serde allowed deserializers to implement interface providers.